File: src/minecraft/net/machinemuse/powersuits/client/render/modelspec/RenderPart.scala
package net.machinemuse.powersuits.client.render.modelspec

import net.machinemuse.general.geometry.Colour
import net.minecraft.client.Minecraft
import net.machinemuse.powersuits.client.render.item.ArmorModel
import net.minecraft.nbt.NBTTagCompound
import net.machinemuse.utils.render.Render
import net.minecraft.client.model.{ModelBase, ModelRenderer}
import net.machinemuse.general.NBTTagAccessor
import org.lwjgl.opengl.GL11._
import org.lwjgl.opengl.GL11

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 4:16 AM, 29/04/13
 */

class RenderPart(base: ModelBase) extends ModelRenderer(base) {
  override def render(par1: Float) {
    val renderSpec = ArmorModel.instance.renderSpec
    import scala.collection.JavaConverters._
    val colours = renderSpec.getIntArray("colours")

    for (nbt <- NBTTagAccessor.getValues(renderSpec).asScala) {
      ModelRegistry.getPart(nbt).map(part => {
        if (part.slot == ArmorModel.instance.visible && part.morph.apply(ArmorModel.instance) == this) {
          withMaybeGlow(part, nbt) {
            Render.withPushedMatrix {
              Render pure {
                GL11.glScaled(par1, par1, par1)
                try {
                  Minecraft.getMinecraft.renderEngine.bindTexture(part.getTexture(nbt))
                  applyTransform
                  val ix = part.getColourIndex(nbt)
                  if (ix < colours.size) {
                    Colour.doGLByInt(colours(ix))
                  }
                  part.modelSpec.applyOffsetAndRotation // not yet implemented
                  part.modelSpec.model.renderPart(part.partName)
                  Colour.WHITE.doGL()
                } catch {
                  case e: Throwable =>
                }
              }
            }
          }.run()
        }
      })
    }
  }

  val degrad = 180F / Math.PI.asInstanceOf[Float]

  def withMaybeGlow(part: ModelPartSpec, nbt: NBTTagCompound) = if (part.getGlow(nbt)) {
    a: Render[_] => Render withGlow a
  } else {
    a: Render[_] => a
  }

  def applyTransform {
    glTranslatef(rotationPointX, rotationPointY, rotationPointZ)
    glRotatef(rotateAngleZ * degrad, 0.0F, 0.0F, 1.0F)
    glRotatef(rotateAngleY * degrad, 0.0F, 1.0F, 0.0F)
    glRotatef(rotateAngleX * degrad + 180, 1.0F, 0.0F, 0.0F)
    glTranslated(field_82906_o, field_82907_q - 26, field_82908_p)

  }
}

File: src/main/java/net/machinemuse/powersuits/client/render/item/armor/RenderPart.java
package net.machinemuse.powersuits.client.render.item.armor;

import net.machinemuse.numina.client.render.RenderState;
import net.machinemuse.numina.client.render.modelspec.ModelPartSpec;
import net.machinemuse.numina.client.render.modelspec.ModelRegistry;
import net.machinemuse.numina.client.render.modelspec.PartSpecBase;
import net.machinemuse.numina.constants.ModelSpecTags;
import net.machinemuse.numina.math.Colour;
import net.machinemuse.numina.nbt.NBTTagAccessor;
import net.machinemuse.powersuits.client.model.item.ArmorModelInstance;
import net.machinemuse.powersuits.client.model.item.HighPolyArmor;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.BufferBuilder;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.entity.model.ModelBase;
import net.minecraft.client.renderer.entity.model.ModelRenderer;
import net.minecraft.client.renderer.model.BakedQuad;
import net.minecraft.client.renderer.texture.TextureMap;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.client.ForgeHooksClient;
import org.lwjgl.opengl.GL11;

import java.util.List;

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 4:16 AM, 29/04/13
 * <p>
 * Ported to Java by lehjr on 11/6/16.
 */
@OnlyIn(Dist.CLIENT)
public class RenderPart extends ModelRenderer {
    ModelRenderer parent;

    public RenderPart(ModelBase base, ModelRenderer parent) {
        super(base);
        this.parent = parent;
    }

    @Override
    public void render(float scale) {
        NBTTagCompound renderSpec = ((HighPolyArmor) (ArmorModelInstance.getInstance())).getRenderSpec();
        if (renderSpec == null)
            return;

        int[] colours = renderSpec.getIntArray(ModelSpecTags.TAG_COLOURS);
        if (colours.length == 0)
            colours = new int[]{Colour.WHITE.getInt()};

        int partColor;
        for (NBTTagCompound nbt : NBTTagAccessor.getValues(renderSpec)) {
            PartSpecBase part = ModelRegistry.getInstance().getPart(nbt);
            if (part != null && part instanceof ModelPartSpec) {
                if (part.getBinding().getSlot() == ((HighPolyArmor) (ArmorModelInstance.getInstance())).getVisibleSection()
                        && part.getBinding().getTarget().apply(ArmorModelInstance.getInstance()) == parent) {
                    List<BakedQuad> quadList = ((ModelPartSpec) part).getQuads();
                    if (!quadList.isEmpty()) {
                        int ix = part.getColourIndex(nbt);

                        // checks the range of the index to avoid errors OpenGL or crashing
                        if (ix < colours.length && ix >= 0) partColor = colours[ix];
                        else partColor = Colour.WHITE.getInt();

                        // GLOW stuff on
                        if (((ModelPartSpec) part).getGlow(nbt))
                            RenderState.glowOn();

                        GlStateManager.pushMatrix();
                        GlStateManager.scalef(scale, scale, scale);
                        applyTransform();
                        Minecraft.getInstance().textureManager.bindTexture(TextureMap.LOCATION_BLOCKS_TEXTURE);
                        Tessellator tess = Tessellator.getInstance();
                        BufferBuilder buffer = tess.getBuffer();
                        buffer.begin(GL11.GL_QUADS, DefaultVertexFormats.ITEM);

                        for (BakedQuad quad : ((ModelPartSpec) part).getQuads()) {
                            buffer.addVertexData(quad.getVertexData());
                            ForgeHooksClient.putQuadColor(buffer, quad, partColor);
                        }
                        tess.draw();

                        GlStateManager.popMatrix();

                        //Glow stuff off
                        if (((ModelPartSpec) part).getGlow(nbt))
                            RenderState.glowOff();
                    }
                }
            }
        }
    }

    private void applyTransform() {
//        float degrad = (float) (180F / Math.PI);
//        GL11.glTranslatef(rotationPointX, rotationPointY, rotationPointZ);
//        GL11.glRotatef(rotateAngleZ * degrad, 0.0F, 0.0F, 1.0F);
//        GL11.glRotatef(rotateAngleY * degrad, 0.0F, 1.0F, 0.0F);
//        GL11.glRotatef(rotateAngleX * degrad, 1.0F, 0.0F, 0.0F);
        GlStateManager.rotatef(180, 1.0F, 0.0F, 0.0F);
        GlStateManager.translatef(offsetX, offsetY - 26, offsetZ);
    }
}